Improve the 'Writing custom documentation' section

Relintai opened the pull request:
I improved the 'Writing custom documentation' section of the Custom modules in C++ page.

I made it a continuation of the tutorial from the previous sections. I also made sure to more explicitly explain, and show how doctool should be used. 

Also this corrects an error, because the current version tells you about `godot/doc/classes/`, but, on godot-master that path should be `godot/modules/summator/doc_classes/`.

Xrayez commented:
Knowing about `doc/classes` fallback mechanism, I also want to share a hack with you. 🙂 

If you have lots of classes for which you want to generate docs, having to list them all in `get_doc_classes()` manually can be a lot of work (you may also miss some during the process). I've even made attempts to solve this in godotengine/godot#33037 but then realized it won't work for when docs need to be generated initially.

Knowing this, you can do a hack to get a list of generated classes as reported by `git status`, namely `Untracked` files:

```
Untracked files:
  (use "git add <file>..." to include in what will be committed)

        doc/classes/MyClass2D.xml
        doc/classes/MyClass4D.xml
        doc/classes/MyClass5D.xml
        doc/classes/MyClass6D.xml
 ...  
```

As formatted it is, you can copy the diff list to the `get_doc_classes()`, and then recompile the engine the second time.

Relintai commented:
Thanks, I'll add these to the pr, when I'll have some time!

Relintai commented:
I added everything to the pr. Sorry, it took a a lot more time to get to it than I originally expected.

I reworked section 2, because I felt that the fallback mechanic needs to be articulated (it would be too complex to explain as just a note), also I added the `--no-docbase` option to section 3 as a tip.

akien-mga commented on development/cpp/custom_modules_in_cpp.rst:
```diff
@@ -449,54 +449,94 @@ There are several steps in order to setup custom docs for the module:
+The ``get_doc_path()`` function is used by the build system to determine
+the location of the docs. In this case, they will be located in the 
+``modules/summator/doc_classes`` directory. If you don't define this,
+the doc path for you module will fall back to the main ``doc/classes`` 
```
```suggestion
the doc path for your module will fall back to the main ``doc/classes`` 
```

akien-mga commented on development/cpp/custom_modules_in_cpp.rst:
```diff
@@ -449,54 +449,94 @@ There are several steps in order to setup custom docs for the module:
+.. tip::
+
+    You can use git to check if you have missed some of your classes by checking the 
+    untracked files with ``git status``. For exmample::
```
```suggestion
    untracked files with ``git status``. For example::
```

akien-mga commented on development/cpp/custom_modules_in_cpp.rst:
```diff
@@ -449,54 +449,94 @@ There are several steps in order to setup custom docs for the module:
+
+3. Now we can generate the documentation:
+
+We can do this via running godot's doctool e.g. ``godot --doctool <path>``,
```
```suggestion
We can do this via running Godot's doctool, i.e. ``godot --doctool <path>``,
```

akien-mga commented on development/cpp/custom_modules_in_cpp.rst:
```diff
@@ -449,54 +449,94 @@ There are several steps in order to setup custom docs for the module:
    ::
 
-      godot --doctool <path>
+      user@host:~/godot$ ./bin/godot* --doctool .
```
That's not going to work if you have more than one `godot*` file in `./bin`.

Xrayez replied:
I'd personally expect the documentation to assume `godot` is at the PATH or so, to simplify this for the sake of example/readability.

Relintai replied:
The problem with this one is the section directly before this one, as we have explanations about building modules as shared libraries (near line 391):

```
...

Once compiled, we should end up with a ``bin`` directory containing both the
``godot*`` binary and our ``libsummator*.so``. However given the .so is not in
a standard directory (like ``/usr/lib``), we have to help our binary find it
during runtime with the ``LD_LIBRARY_PATH`` environ variable:

::

    user@host:~/godot$ export LD_LIBRARY_PATH=`pwd`/bin/
    user@host:~/godot$ ./bin/godot*

... bit more explanation ...

Writing custom documentation
----------------------------
```

I have an idea that *might* be the best of both worlds. I'll push it soon, along with all the other corrections.

Also I personally don't think the executable should be in the PATH in this example, because we build it a section earlier.

akien-mga commented on development/cpp/custom_modules_in_cpp.rst:
```diff
@@ -449,54 +449,94 @@ There are several steps in order to setup custom docs for the module:
+.. note::
 
-.. code-block:: console
+    If you want, you can prevent the core classes to be generated with the 
```
```suggestion
    If you want, you can prevent the core classes from being generated with the 
```

akien-mga replied:
But I'd straight out remove this note, it's not relevant for this use case.
What it does it that it doesn't dump documentation for Variants, but it still includes everything else from the API, so:
- Its actual use case is dubious
- It's not relevant for writing module docs

Xrayez replied:
Last time I tested `--no-docbase` will also remove all base docs as well, polluting the `git status` with deleted classes, but this something which should be fixed IMO.

Relintai replied:
I removed this one.

akien-mga commented on development/cpp/custom_modules_in_cpp.rst:
```diff
@@ -449,54 +449,94 @@ There are several steps in order to setup custom docs for the module:
-4. Get generated doc file from ``godot/doc/classes/ClassName.xml``
+Now if you go to the ``godot/modules/summator/doc_classes`` folder, you will see, 
+that now we have a ``Summator.xml`` inside it, along with every class, that 
+you added with your ``get_doc_classes`` function.
```
```suggestion
Now if you go to the ``godot/modules/summator/doc_classes`` folder, you will see 
that it contains a ``Summator.xml`` file, or any other classes, that you referenced
in your ``get_doc_classes`` function.
```

akien-mga commented on development/cpp/custom_modules_in_cpp.rst:
```diff
@@ -449,54 +449,94 @@ There are several steps in order to setup custom docs for the module:
+you added with your ``get_doc_classes`` function.
 
-5. Copy this file to ``doc_classes``, optionally edit it, then compile the engine.
+Now you can edit this file, and recompile the engine.
```
```suggestion
Edit the file(s) following :ref:`doc_updating_the_class_reference` and recompile the engine.
```

akien-mga commented on development/cpp/custom_modules_in_cpp.rst:
```diff
@@ -449,54 +449,94 @@ There are several steps in order to setup custom docs for the module:
 
 In order to keep documentation up-to-date, all you'll have to do is simply modify
-one of the ``ClassName.xml`` files and recompile the engine from now on.
+one of the xml files and recompile the engine from now on.
```
```suggestion
one of the XML files and recompile the engine from now on.
```

Relintai commented:
Pushed up the new version. 

Apart from the typo fixes, In 'Improving the build system for development' I also touched 2 lines a little bit.

About the `godot*` commands, I opted to change both occurrences to `<godot_binary>`, as the binary's name changes for every platform and target.

I also thought about changing the current directory to `bin`, but I think that would easily lead to confusion when moving  between sections. Al least as a bonus everything can still be easily copy pasted.

Note:
I think that the 'Improving the build system for development' section could be moved to the end of the page, above the 'Summing up' section.
Actually, both it's name, and it's content makes it fit there perfectly (for example it only works on linux). This would makes the document flow a bit better.
(I included this first, but realized that it will make reviewing annoying, so I removed it. I will open a separate pr for it.)

mhilbrunner commented:
Hey! Sorry it took a bit to get back to this.

Thanks for the PR, it looks good to merge to me - could you fix the conflict? :)

Relintai commented:
Done

mhilbrunner commented:
Agreed :)

Thank you for this, and again, sorry it took a while!
